training/creating_dataset.py

```diff
@@ -15,117 +13,106 @@
-    return inner_move_values_level_up
+    if characteristics is None:
+        characteristics = ["abilities", "ac", "hp"]
 
+    data = []
 
-def get_subcolumn(book: pd.DataFrame, subcolumn_path: str) -> pd.DataFrame:
-    """
-    Gets subcolumn of given DataFrame according to given path
-    :param book: DataFrame with all data from book(s)
-    :param subcolumn_path: path to subcolumn
-    :return: chosen subcolumn as DataFrame
-    """
-    subcol = deepcopy(book)
-    for col in subcolumn_path.split("/"):
-        subcol = pd.DataFrame(data=unpack_column(subcol, col))
+    for path in paths_to_books:
+        is_path_correct(path)
+        with open(path) as file:
+            # loading json strings from files
+            d = [json.loads(line) for line in file]
+            # getting only system column (all characteristics are there)
+            # getting only npc - monsters
+            data += [line.get("system") for line in d if line.get("type") == "npc"]
+            file.close()
 
-    return subcol
+    bestiary = pd.json_normalize(data)
 
+    columns = []
+    for characteristic in set(characteristics + ["book", "level"]):
+        # different way of loading data: abilities and saves is a group of characteristics
+        if characteristic in ["abilities", "saves"]:
+            subcolumns = bestiary.filter(regex=characteristic)
 
-def load_subcolumn_as_value(
-    column_name: str, original_column_name: str = "value"
-) -> callable:
-    """
-    Returns a function that creates DataFrame with chosen value_name of given DataFrame
-    and changes column name to chosen one
-    :param column_name: name that should be given to result dataframe column
-    :param original_column_name: name of subcolumn which value will be returned
-    :return: function to create Dataframe with chosen column name and values from a chosen subcolumn of current column
-    """
+            # in case of saves we have to filter to extract only useful columns
+            if characteristic == "saves":
+                subcolumns = subcolumns.filter(regex="value")
 
-    def subcolumn_as_value(df: pd.DataFrame) -> pd.DataFrame:
-        result_df = pd.DataFrame(data=df[original_column_name])
-        result_df.columns = [column_name]
-        return result_df
+            # renaming columns to simpler names of characteristics
+            subcolumns = subcolumns.rename(
+                columns={col: col.split(".")[-2] for col in subcolumns.columns}
+            )
+            columns.append(subcolumns)
+            continue
+        # renaming the column to a simpler name of characteristic
+        column = bestiary[CHARACTERISTICS_COLUMNS.get(characteristic)].rename(
+            characteristic
+        )
+        columns.append(column)
 
-    return subcolumn_as_value
+    df = pd.concat(columns, axis=1)
+    # there is an option to load all abilities and also each of them separately - get rid of duplicates
+    df = df.loc[:, ~df.columns.duplicated()]
```

Comment:
Czemu w ogóle może dojść do duplikacji? Przecież explicite zakładamy płaską listę konkretnych cech

Reply:
Rzecz w tym, że tutaj nie ma użytkownika. W sensie użytkownikiem jesteśmy my, jako backend projektujący model. Mamy pełną wiedzę i kontrolę, a grupa cech to jest zmienna ciut wyżej.

Ogólnie, kod MLowy do treningu modelu jest specyficzny, bo to jest kawał imperatywnego kodu z konkretnymi założeniami wbudowanymi w niego. Różni się od typowego user-facing kodu OOP, bo tutaj mamy pełną kontrolę, wiedzę i też prawie nigdy nie potrzebujemy takiej ogólności działania.

Reply:
ok...
to ma się pozbyć `saves` i `abilities`?